### PR TITLE
Add simple Identity Map

### PR DESCRIPTION
**Pull request by richmolj**
* When persisted, model instances are registered with an in-memory
store. When deleted, they are removed from the store and the new `stale`
property becomes `true`.
* Model attributes can now be synced across instances. If assigning an
attribute, that attribute will not be synced until save. After save, all
instances of that Model will sync attributes.
* In order to sync attributes, we listen for updates from the central
store. This means to be garbage collected the listener must be removed.
You can do this by calling `model.unlisten()`. Framework-specific
libraries will ensure `unlisten()` is called automatically when the
framework no longer references the instance.
* When associations return related objects, they remove anything that
was deleted from the store. This way `await book.destroy()` will affect
the `author.books` relationship.

Because the behavior of modifying instances out from under you can be
seen as an implicit (and possibly undesired) side-effect, this behavior
is opt-in:

```ts
// ApplicationRecord
static sync = true
```

### CODE REVIEW

**Comment by richmolj** [comment]
@wadetandy - new PR basing this on eventbus, which removes attributeOverrides. FYI was able to make this refactor without changing any of the idmap integration tests.

File: src/attribute.ts [hunk 1]
```diff
@@ -66,7 +66,7 @@ export class Attribute<T = any> {
   // The model calls this getter
   getter(context: JSORMBase): any {
     const privateContext: any = context
-    return privateContext._attributes[this.name]
+    return privateContext.attributes[this.name]
```
Comment (wadetandy) [inline src/attribute.ts]: Should be able to remove the privateContext but here if moving to attributes.

File: src/model.ts [hunk 2]
```diff
@@ -214,6 +216,10 @@ export class JSORMBase {
       this._middlewareStack = new MiddlewareStack()
     }
 
+    if (!this._IDMap) {
```
Comment (wadetandy) [inline src/model.ts]: Shouldn’t assign unless `sync==true`?

File: src/model.ts [hunk 3]
```diff
@@ -405,10 +421,51 @@ export class JSORMBase {
...
+    if (this._onStoreChange) return this._onStoreChange
+    this._onStoreChange =  (_event: any, attrs: any) => {
+      Object.keys(attrs).forEach((k) => {
+        if (this[k] !== attrs[k]) this[k] = attrs[k]
+      })
+    }
+    return this._onStoreChange
+  }
+
+  unlisten() {
```
Comment (wadetandy) [inline src/model.ts]: Would mark these two as `void` return types. Could see this accidentally evolving somehow.

File: test/integration/id-map.test.ts [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+      expect(author2.firstName).to.not.eq('updated')
+
+      await author1.save()
+
+      // now synced after save
+      expect(author2.firstName).to.eq('updated')
+
+      // now back to unsynced
+      author1.firstName = 'updated again'
+      expect(author2.firstName).to.eq('updated')
```
Comment (wadetandy) [inline test/integration/id-map.test.ts]: Don’t have to do it for every case of this, but I would recommend adding one or two smoke tests around calling `unlisten()` not causing further changes, both if the unlistened model saves itself and if another saves.

File: src/event-bus.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+  (key: string, context: any, value: any): void
+}
+
+export interface IEventBus {
+  addEventListener: IAddListener
+  removeEventListener: IRemoveListener
+  dispatch: IDispatch
+}
+
+const EventBus: IEventBus = require('eventbusjs')
```
Comment (wadetandy) [inline src/event-bus.ts]: Fine with this lib and seems like it does everything we need so whatever, but did you consider using the node `EventEmitter` API instead? Would make it more standard for future contributors and potentially could even get a browser-native version eventually. There's a few browser implementations of this that we could pull in as a hard dependency (or wrap this lib in a small shim to make the API match).

Comment (richmolj) [reply src/event-bus.ts]: I'm not opposed to this, but our usage is so simplistic I'm not sure it's work the overhead of changing. We could always refactor.

**Pull request merged**